src/main/java/org/jenkinsci/plugins/credentialsbinding/impl/BindingStep.java

```diff
@@ -75,20 +80,41 @@
         return bindings;
     }
 
-    public static final class Execution extends AbstractStepExecutionImpl {
+    @Override
+    public StepExecution start(StepContext context) throws Exception {
+        return new Execution(this, context);
+    }
+
+    public static final class Execution extends StepExecution {
 
         private static final long serialVersionUID = 1;
 
-        @Inject(optional=true) private transient BindingStep step;
-        @StepContextParameter private transient Run<?,?> run;
-        @StepContextParameter private transient FilePath workspace;
-        @StepContextParameter private transient Launcher launcher;
-        @StepContextParameter private transient TaskListener listener;
+        private transient BindingStep step;
+
+        public Execution(@Nonnull BindingStep step, StepContext context) {
+            super(context);
+            this.step = step;
+        }
 
         @Override public boolean start() throws Exception {
+            Run<?,?> run = getContext().get(Run.class);
+            if (run == null) {
+                throw new MissingContextVariableException(Run.class);
+            }
+
+            FilePath workspace = getContext().get(FilePath.class);
+            Launcher launcher = getContext().get(Launcher.class);
+            TaskListener listener = getContext().get(TaskListener.class);
```

Comment:
why not fail quickly here as well since you claim to require a `TaskListener` below in `getRequiredContext`?